src/metatrain/experimental/phace/trainer.py

```diff
@@ -0,0 +1,644 @@
+            if not model.has_new_targets:
+                optimizer.load_state_dict(self.optimizer_state_dict)
+
+        # Create a learning rate scheduler
+        lr_scheduler = get_scheduler(optimizer, self.hypers, len(train_dataloader))
+
+        if self.scheduler_state_dict is not None:
+            # same as the optimizer, try to load the scheduler state dict
+            if not model.has_new_targets:
+                lr_scheduler.load_state_dict(self.scheduler_state_dict)
+
+        # per-atom targets:
+        per_structure_targets = self.hypers["per_structure_targets"]
+
+        # Log the initial learning rate:
+        old_lr = optimizer.param_groups[0]["lr"]
+        logging.info(f"Initial learning rate: {old_lr}")
+
+        start_epoch = 0 if self.epoch is None else self.epoch + 1
+
+        # Train the model:
+        if self.best_metric is None:
+            self.best_metric = float("inf")
+        logging.info("Starting training")
+        epoch = start_epoch
+        for epoch in range(start_epoch, start_epoch + self.hypers["num_epochs"]):
+            if is_distributed:
+                for train_sampler in train_samplers:
+                    train_sampler.set_epoch(epoch)
+
+            train_rmse_calculator = RMSEAccumulator(self.hypers["log_separate_blocks"])
+            val_rmse_calculator = RMSEAccumulator(self.hypers["log_separate_blocks"])
+            if self.hypers["log_mae"]:
+                train_mae_calculator = MAEAccumulator(
+                    self.hypers["log_separate_blocks"]
+                )
+                val_mae_calculator = MAEAccumulator(self.hypers["log_separate_blocks"])
+
+            train_loss = 0.0
+            for batch in train_dataloader:
+                optimizer.zero_grad()
+
+                systems, targets, extra_data = unpack_batch(batch)
+                systems, targets, extra_data = batch_to(
+                    systems, targets, extra_data, dtype=dtype, device=device
+                )
+
+                predictions = model(
+                    systems, _get_requested_outputs(targets, dataset_info.targets)
+                )
+
+                # average by the number of atoms
+                predictions = average_by_num_atoms(
+                    predictions, systems, per_structure_targets
+                )
+                targets = average_by_num_atoms(targets, systems, per_structure_targets)
+
+                train_loss_batch = loss_fn(predictions, targets, extra_data)
+
+                train_loss_batch.backward()
+
+                # In distributed training, manually average gradients across processes
+                # instead of using DDP to avoid gradient bucketing breaking the graph
+                if is_distributed:
+                    for param in model.parameters():
+                        if param.grad is not None:
+                            torch.distributed.all_reduce(param.grad)
+                            param.grad /= world_size
```

Comment:
Small question: why this and not the below as in PET? Is there any real difference?
```suggestion
                if is_distributed:
                    # make sure all parameters contribute to the gradient calculation to
                    # make torch DDP happy
                    for param in model.parameters():
                        train_loss_batch += 0.0 * param.sum()
```

Reply:
We're not using distributed data parallel from torch here because it generates graph breaks for torch-compile due to the bucketing communication it implements (you can ask LLMs for details)